File: src/main/resources/static/js/app/services/articleService.js
/**
 * Service for articles functions
 */

webShopApp.factory('articlesFactory', function($http)
{	
	let factory = {};
	
	factory.add = function(article) 
	{
		return $http.post('/api/articles', article);
	};
	
	factory.update = function(article) 
	{
		return $http.put('/api/articles', article);
	};
	
	factory.delete = function(id) 
	{
		return $http.delete('/api/articles/' + id);
	};
	
	factory.getArticles = function() 
	{
		return $http.get('/api/articles');
	};
	
	factory.getArticlesOnDiscount = function() 
	{
		return $http.get('/api/articles/discount');
	};
	
	factory.getArticlesByCategory = function(categoryId) 
	{
		return $http.get('/api/articles/category/' + categoryId);
	};
	
	factory.getSortedArticles = function(sortType) 
	{
		return $http.get('/api/articles/sort/' + sortType);
	};
	
	factory.searchArticles = function(name, description, priceFrom, priceTo) 
	{
		let path = '/api/articles/search';
		let firstParameterPassed = false;
		
		if (name && name.length > 0)
		{
			path = path.concat('?name=' + name);
			firstParameterPassed = true;
		}
		
		if (description && description.length > 0)
		{
			if (!firstParameterPassed)
			{
				path = path.concat('?description=' + description);
				firstParameterPassed = true;
			}
			else
			{
				path = path.concat('&description=' + description);
			}
		}
		
		if (priceFrom && !isNaN(priceFrom))
		{
			if (!firstParameterPassed)
			{
				path = path.concat('?priceFrom=' + priceFrom);
				firstParameterPassed = true;
			}
			else
			{
				path = path.concat('&priceFrom=' + priceFrom);
			}
		}
		
		if (priceTo && !isNaN(priceTo))
		{
			if (!firstParameterPassed)
			{
				path = path.concat('?priceTo=' + priceTo);			
			}
			else
			{
				path = path.concat('&priceTo=' + priceTo);
			}
		}
		
		return $http.get(path);
	};
	
	return factory;
});

File: src/main/java/org/web/onlineshop/service/ArticleService.java
package org.web.onlineshop.service;

import java.util.ArrayList;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.web.onlineshop.exceptions.ArticleAlreadyExistsException;
import org.web.onlineshop.exceptions.ArticleNotExistsException;
import org.web.onlineshop.model.Article;
import org.web.onlineshop.model.ArticleCategory;
import org.web.onlineshop.repository.ArticleRepository;
import org.web.onlineshop.util.ArticleSortType;

@Service
public class ArticleService 
{
	@Autowired
	private ArticleRepository articleRepository;
	
	public List<Article> findAll()
	{
		return this.articleRepository.findAll();
	}
	
	public List<Article> findAllOnDiscount()
	{
		return this.articleRepository.findByDiscountNotNullAndDiscountGreaterThan(0);
	}
	
	public List<Article> findByCategory(ArticleCategory category)
	{
		return this.articleRepository.findByCategory(category);
	}
	
	public List<Article> findAllSorted(ArticleSortType sortType)
	{
		if (sortType == ArticleSortType.NAME_ASC)
		{
			return this.articleRepository.findAllByOrderByNameAsc();
		}
		else if (sortType == ArticleSortType.NAME_DESC)
		{
			return this.articleRepository.findAllByOrderByNameDesc();
		}
		else if (sortType == ArticleSortType.PRICE_ASC)
		{
			return this.articleRepository.findAllByOrderByPriceAsc();
		}
		else if (sortType == ArticleSortType.PRICE_DESC)
		{
			return this.articleRepository.findAllByOrderByPriceDesc();
		}
		else
		{
			return new ArrayList<>();
		}
	}
	
	public List<Article> search(String name, String description, Double priceFrom, Double priceTo)
	{
		List<Article> articles = (name != null) ? this.articleRepository.findByNameIgnoreCaseContaining(name) : this.articleRepository.findAll();
	
		if (description != null)
		{
			articles.retainAll(this.articleRepository.findByDescriptionIgnoreCaseContaining(description));
		}
		
		if (priceFrom != null)
		{
			articles.retainAll(this.articleRepository.findByPriceGreaterThanEqual(priceFrom));
		}
		
		if (priceTo != null)
		{
			articles.retainAll(this.articleRepository.findByPriceLessThanEqual(priceTo));
		}
		
		return articles;
	}
	
	public Article save(Article article)
	{
		if (!this.articleRepository.findByName(article.getName()).isEmpty())
		{
			throw new ArticleAlreadyExistsException(article.getName());
		}
		return this.articleRepository.save(article);
	}
	
	public Article update(Article article)
	{
		if (this.articleRepository.existsById(article.getId()))
		{
			return this.articleRepository.save(article);			
		}
		else
		{
			throw new ArticleNotExistsException(article.getId());
		}
	}
	
	public void delete(Long id)
	{
		this.articleRepository.deleteById(id);
	}
}